File: C++/20Company/10Bytedance_US/1249MinimumRemovetoMakeValidParentheses/solwithStack.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        /* 
                   i
            s   = "lee(t(c)o)de)"
            st: base|[ )
        */
        stack<int> st;
        for(int i=0;i<s.size();){
            if(s[i]=='('){
                st.push(i);
                i++;
            }
            else if(s[i]==')'){
                if(st.empty()){
                    s.erase(i,1);
                }
                else{
                    st.pop();
                    i++;
                }
            }
            else{
                i++;
            }
        }
        while(!st.empty()){
            s.erase(st.top(), 1);
            st.pop();
        }
        return s;
    }
};

File: Java/08Stack&Queue/src/lc20ValidParentheses/SolWithStack.java
package lc20ValidParentheses;
import java.util.Stack;
import java.util.Scanner;

public class SolWithStack {
    public static void main(String[] args){
        Scanner scan = new Scanner(System.in);
        Solution S1 = new Solution();
        while(true){
            System.out.println("Enter string s:");
            String s = scan.nextLine();
            if(s.equals("q")) break;
            System.out.println("Res: "+S1.isValid(s));
        }
    }
    private static class Solution{
        public boolean isValid(String s) {
            /*
                Input: s = "()[]{}" Output: true
                {[
             */
            Stack<Character> st = new Stack<>();
            st.push(s.charAt(0));
            for(int i=1;i<s.length();++i){
                char cur = s.charAt(i);
                if(st.empty()){
                    st.push(cur);
                    continue;
                }
                char tmp = st.peek();
                if((tmp=='('&&cur==')')||(tmp=='['&&cur==']')||(tmp=='{'&&cur=='}')){
                    st.pop();
                }
                else{
                    st.push(cur);
                }
            }
            return st.empty();
        }
    }
}
